### PR TITLE
BGDIINF_SB-2890 : fix 3d tests

### PR DESCRIPTION
**Pull request by pakb**
URL param parsing : processing projection before position, so that when we (cold) start the app with &3d=true, it correctly transition from the default center to the same one but re-projected in Mercator

Moving all "position" related reprojection from the store plugin to the store file (no need to separate it, as it is the responsibility of the store file to handle its own stuff)

In Cypress commands.js : only hard coding the center to the old value when 3D flag is not present (otherwise a center expressed in LV95 is added to the URL while it should function in Mercator when in 3D)

[Test link](https://sys-map.dev.bgdi.ch/preview/bug_bgdiinf_sb-2890_fix_3d_tests/index.html)

### CODE REVIEW

File: src/router/legacyPermalinkManagement.routerPlugin.js [hunk 1]
```diff
@@ -111,8 +111,16 @@ const handleLegacyParam = (
...
 
-        // if no special work to do, we just copy past legacy params to the new viewer
+        case '3d':
+            // if the 3d flag is given without being placed behind the hash (meaning at startup),
+            // we need to make sure the projection is set to Mercator, otherwise the startup sequence
+            // ends up not working properly and center the view wrongly
+            if ((typeof legacyValue === 'string' && legacyValue === 'true') || legacyValue) {
+                newQuery['sr'] = WEBMERCATOR.epsgNumber
+            }
+        // no break so we go into default too
```
Comment (ltshb) [inline src/router/legacyPermalinkManagement.routerPlugin.js:121]: `3d` is not a legacy parameter and we should never have it before the `#` of I'm not misteaken ?

File: src/store/modules/position.store.js [hunk 2]
```diff
@@ -264,6 +259,37 @@ const actions = {
             )
         }
         if (matchingProjection) {
+            const oldProjection = state.projection
```
Comment (ltshb) [inline src/store/modules/position.store.js:262]: Should we not check if the projection isalready correct before doing all this work ?

```
if (matchingProjection && matchingProjection.epsg != oldProjection.epsg) {
...
}
```

File: src/store/modules/position.store.js [hunk 3]
```diff
@@ -264,6 +259,37 @@ const actions = {
...
+            if (
+                oldProjection instanceof CustomCoordinateSystem &&
+                matchingProjection instanceof StandardCoordinateSystem
+            ) {
+                commit('setZoom', oldProjection.transformCustomZoomLevelToStandard(state.zoom))
+            }
+            if (
+                oldProjection instanceof CustomCoordinateSystem &&
+                matchingProjection instanceof CustomCoordinateSystem &&
+                oldProjection.epsg !== matchingProjection.epsg
```
Comment (ltshb) [inline src/store/modules/position.store.js:282]: See comment above, I guess if the projection did not change we don't need to change the center nor the zoom ?

File: src/store/plugins/reproject-everything-on-projection-change.plugin.js [hunk 4]
```diff
@@ -32,9 +32,6 @@ const reprojectEverythingOnProjectionChangePlugin = (store) => {
                 const reprojectCoordinates = (coordinates = []) =>
                     proj4(oldProjection.epsg, newProjection.epsg, coordinates)
 
-                // reprojecting the center of the map
-                store.dispatch('setCenter', reprojectCoordinates(state.position.center))
```
Comment (ltshb) [inline src/store/plugins/reproject-everything-on-projection-change.plugin.js:36]: I would suggest to rename this plugin as it only reproject selected features and not everything.

File: tests/e2e-cypress/support/commands.js [hunk 5]
```diff
@@ -132,7 +132,7 @@ Cypress.Commands.add(
         if (!queryParams.hasOwnProperty('lang')) {
             queryParams.lang = 'en'
         }
-        if (!queryParams.hasOwnProperty('center')) {
+        if (!queryParams.hasOwnProperty('center') && !queryParams.hasOwnProperty('3d')) {
```
Comment (ltshb) [inline tests/e2e-cypress/support/commands.js:135]: Here I would also not add the center in legacy url, if `withHash` == false

File: src/router/legacyPermalinkManagement.routerPlugin.js (hunk 1 above)
Comment (pakb) [reply src/router/legacyPermalinkManagement.routerPlugin.js:121]: In some of our tests, it is.
And even though it's not legacy, I think we should make sure (advanced) users don't have to write this # to make it work.

We could think about having a dedicated file for current URL params than need special care, that would make more sense

File: src/store/modules/position.store.js (hunk 2 above)
Comment (pakb) [reply src/store/modules/position.store.js:262]: might be a good idea, yes

File: src/router/legacyPermalinkManagement.routerPlugin.js (hunk 1 above)
Comment (ltshb) [reply src/router/legacyPermalinkManagement.routerPlugin.js:121]: As discussed we do this later on

**Review by ltshb — APPROVED** [review]

**Pull request merged**